plutus-core/generators/Language/PlutusCore/Generators/NEAT/Type.hs

```diff
@@ -46,213 +62,374 @@ data TypeBuiltinG
   where
     k' = KindArrow () (KindArrow () k (Type ())) (KindArrow () k (Type ()))
-toType ns (Type _) (TyForallG k ty) = do
-  ns' <- extendTyNameState ns
-  TyForall () (tynameOf ns' FZ) k <$> toType ns' (Type ()) ty
-toType _ _ (TyBuiltinG tyBuiltin) =
-  return (TyBuiltin () (toTypeBuiltin tyBuiltin))
-toType ns (KindArrow _ k1 k2) (TyLamG ty) = do
-  ns' <- extendTyNameState ns
-  TyLam () (tynameOf ns' FZ) k1 <$> toType ns' k2 ty
-toType ns k2 (TyAppG ty1 ty2 k1) =
-  TyApp () <$> toType ns k' ty1 <*> toType ns k1 ty2
-  where
-    k' = KindArrow () k1 k2
-toType _ k ty = throwError $ Gen ty k
-
-
--- * Enumerating closed types
-
-type ClosedTypeG = TypeG Z
-
+convertType tns (Type _) (TyForallG k ty) = do
+  tns' <- extTyNameState tns
+  TyForall () (tynameOf tns' FZ) k <$> convertType tns' (Type ()) ty
+convertType _ _ (TyBuiltinG tyBuiltin) =
+  return (TyBuiltin () (convertTypeBuiltin tyBuiltin))
+convertType tns (KindArrow _ k1 k2) (TyLamG ty) = do
+  tns' <- extTyNameState tns
+  TyLam () (tynameOf tns' FZ) k1 <$> convertType tns' k2 ty
+convertType tns k2 (TyAppG ty1 ty2 k1) =
+  TyApp () <$> convertType tns (KindArrow () k1 k2) ty1 <*> convertType tns k1 ty2
+convertType _ k ty = throwError $ BadTypeG k ty
 
 -- |Convert generated closed types to Plutus types.
-toClosedType
+convertClosedType
   :: (MonadQuote m, MonadError GenError m)
   => Stream.Stream Text.Text
   -> Kind ()
   -> ClosedTypeG
   -> m (Type TyName DefaultUni ())
-toClosedType strs = toType (emptyTyNameState strs)
+convertClosedType tynames = convertType (emptyTyNameState tynames)
+
+
+-- ** Converting terms
+
+-- |Convert (well-typed) generated terms to Plutus terms.
+--
+-- NOTE: Passes an explicit `TyNameState` and `NameState`, instead of using a
+--       State monad, as the type of the `TyNameState` changes throughout the
+--       computation. This could be written using an indexed State monad.
+convertTerm
+  :: (Show tyname, Show name, MonadQuote m, MonadError GenError m)
+  => TyNameState tyname -- ^ Type name environment with fresh name stream
+  -> NameState name     -- ^ Name environment with fresh name stream
+  -> TypeG tyname       -- ^ Type of term below
+  -> TermG tyname name  -- ^ Term to convert
+  -> m (Term TyName Name DefaultUni ())
+convertTerm _tns ns _ty (VarG i) =
+  return (Var () (nameOf ns i))
+convertTerm tns ns (TyFunG ty1 ty2) (LamAbsG tm) = do
+  ns' <- extNameState ns
+  ty1' <- convertType tns (Type ()) ty1
+  LamAbs () (nameOf ns' FZ) ty1' <$> convertTerm tns ns' ty2 tm
+convertTerm tns ns ty2 (ApplyG tm1 tm2 ty1) =
+  Apply () <$> convertTerm tns ns (TyFunG ty1 ty2) tm1 <*> convertTerm tns ns ty1 tm2
+convertTerm tns ns (TyForallG k ty) (TyAbsG tm) = do
+  tns' <- extTyNameState tns
+  TyAbs () (tynameOf tns' FZ) k <$> convertTerm tns' ns ty tm
+convertTerm tns ns _ (TyInstG tm cod ty k) =
+  TyInst () <$> convertTerm tns ns (TyForallG k cod) tm <*> convertType tns k ty
```

Comment:
Do we maybe need to check that the expected type matches with the one that we get by assembling pieces from `TyInstG`?

Reply:
I don't think this is intended to do any checking, the only reason the type is present is because `TermG` doesn't contain all the information needed to make a fully annotated `Term`. I added a note about this above the definition.